Added python section (#250)
OWASP/CheatSheetSeries: cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.md

A description of how to abuse this in PHP is presented in a good [SensePost article](https://www.sensepost.com/blog/2014/revisting-xxe-and-abusing-protocols/) describing a cool PHP based XXE vulnerability that was fixed in Facebook.

# Python

> The following information for XXE injection in python is directly from this [official python documentation page](https://docs.python.org/2/library/xml.html#xml-vulnerabilities).

Reviewer: Given that Python2 will reach its end of life in about three months, I think it would be better to give the Python3 table (as that's why people should be using), with maybe a link to the Python2 table with a note that it's legacy and soon to be unsupported.

Also `Python` should be capitalised.

Author: I have added the main table reference for python 3 . And also link to python 2 with a note as suggested.

> The below table gives an overview of various modules in python used for xml parsing and whether they are vulnearble or not.

Reviewer: `Python` should be capitalised and `XML` should be in block caps for consistency.

`Vulnerable` is spelled incorrectly.

Author: These changes are done

> | kind                      | sax        | etree      | minidom    | pulldom    | xmlrpc     |

Reviewer: This table is broken if you look at the rendered markdown - I think need a blank line before it to render properly. Also, would something like `Attack` or `Attack Type` read better than `kind`?

Author: I've fixed the table now.

|---------------------------|------------|------------|------------|------------|------------|

> | billion laughs            | Vulnerable | Vulnerable | Vulnerable | Vulnerable | Vulnerable |

Reviewer: Please capitalise the attack names at the start of the lines

Author: This is done

| quadratic blowup          | Vulnerable | Vulnerable | Vulnerable | Vulnerable | Vulnerable |

> | external entity expansion | Vulnerable | Safe (1)   | Safe (2)   | Vulnerable | Safe (3)   |

Reviewer: If we don't have the footnotes form the Python docs, the `(1)` and `(2)` should probably be removed.

Author: This is done

| DTD retrieval             | Vulnerable | Safe       | Safe       | Vulnerable | Safe       |
| decompression bomb        | Safe       | Safe       | Safe       | Safe       | Vulnerable |

> Currently there are no explict methods available for above mentioned python modules to mitigate the vulnerablity. Hence sanitizes input prior to processing using the above libraires. 

Reviewer: Capitalisation of `Python`.

Author: Done

Reviewer 2: You appear to have taken this comment incorrectly. Afaict, you changed the other instances of `Python` to `PYTHON` instead of changing this one instance of `python` to `Python` :-(

A description of how to abuse this in PHP is presented in a good [SensePost article](https://www.sensepost.com/blog/2014/revisting-xxe-and-abusing-protocols/) describing a cool PHP based XXE vulnerability that was fixed in Facebook.

# PYTHON

> The following information for XXE injection in PYTHON 3 is directly from this [official PYTHON 3 documentation page](https://docs.python.org/3/library/xml.html#xml-vulnerabilities).

Reviewer 2: Why write `PYTHON` instead of `Python`?

> The following information for XXE injection in PYTHON 3 is directly from this [official PYTHON 3 documentation page](https://docs.python.org/3/library/xml.html#xml-vulnerabilities).

Reviewer 2: I'd probably write `the official` (this applies below as well).

Reviewer 2: The link text should really indicate that it relates to `xml`. (People shouldn't have to read the url which isn't visible to figure this out.)

> The below table gives an overview of various modules in PYTHON 3 used for XML parsing and whether they are vulnerable or not.

Reviewer 2: fwiw `following` would be better than `below`

> > The end of life for Python 2 is expected for January 1st 2020.

Reviewer 2: ```suggestion
> The end of life for Python 2 is expected to be January 1st, 2020.
```

> The following table gives an overview of various modules in Python 3 used for XML parsing and whether they are vulnerable or not.

Reviewer 2: ```suggestion
The following table gives an overview of various modules in Python 3 used for XML parsing and whether or not they are vulnerable.
```

|---------------------------|------------|------------|------------|------------|------------|
| Billion Laughs            | Vulnerable | Vulnerable | Vulnerable | Vulnerable | Vulnerable |
| Quadratic Blowup          | Vulnerable | Vulnerable | Vulnerable | Vulnerable | Vulnerable |
| External Entity Expansion | Safe       | Safe       | Safe       | Safe       | Safe       |
| DTD Retrieval             | Safe       | Safe       | Safe       | Safe       | Safe       |
| Decompression Bomb        | Safe       | Safe       | Safe       | Safe       | Vulnerable |

> For details on PYTHON 2, refer this [official PYTHON 2 documentation page](https://docs.python.org/2/library/xml.html#xml-vulnerabilities). Kindly note that PYTHON 2 End of Life has been announced for January 1st 2020. After that PYTHON 2 will become obsolete and no official support will be available.

Reviewer 2: `Kindly` is an odd word (I can't remember which English variant uses it, but it isn't standard for en-US). I'd suggest `Do` instead, or simply omit the word entirely (as in `Note:`).

Reviewer 2: Note: it's more common to write `the end of life for X` or `X has reached end of life` as opposed to `X end of life` (this isn't about the casing `End of Life`/`end of life`, just the phrasing).

For Python, it hasn't technically reached end of life, but the same general idea applies, as in:
`the end of life for X will/is-expected-to be ...` or `X will reach end of life...`.

> In order to protect your application from the applicable attacks, [2 packages](https://docs.python.org/3/library/xml.html#the-defusedxml-and-defusedexpat-packages) exist in order to help you sanitize your input and protect your application against DDoS and remote attacks.

Reviewer 2: ```suggestion
To protect your application from the applicable attacks, [two packages](https://docs.python.org/3/library/xml.html#the-defusedxml-and-defusedexpat-packages) exist to help you sanitize your input and protect your application against DDoS and remote attacks.
```

> Currently there are no explicit methods available for above mentioned python modules to mitigate the vulnerability. Hence sanitizes input prior to processing using the above libraries. 

Reviewer 2: `for the above-mentioned Python` ... to mitigate `these vulnerabilities`.

Hence `you must sanitize`.

Reviewer 2: Note that the second sentence is ambiguous, it either means `"you should use the above libraries to sanitize your input" and then process it ...` or `"you should use something else to sanitize your input" and then take that sanitized data and feed it to one of these libraries for further processing`.

I know what you're trying to say. But, you've failed to unambiguously state it. And someone should fix this before it's merged.
